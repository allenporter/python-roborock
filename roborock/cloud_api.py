from __future__ import annotations

import asyncio
import base64
import logging
import threading
import typing
import uuid
from asyncio import Lock
from urllib.parse import urlparse

import paho.mqtt.client as mqtt

from .api import KEEPALIVE, RoborockClient
from .containers import DeviceData, UserData
from .exceptions import RoborockException, VacuumError
from .protocol import MessageParser, Utils, md5hex
from .roborock_future import RoborockFuture
from .util import RoborockLoggerAdapter

if typing.TYPE_CHECKING:
    pass
_LOGGER = logging.getLogger(__name__)
CONNECT_REQUEST_ID = 0
DISCONNECT_REQUEST_ID = 1


class MqttClient(mqtt.Client):
    """Implementation of the MQTT client.

    This class exists to manage internals of the client.
    """

    _thread: threading.Thread
    _client_id: str

    def __init__(self, logger: RoborockLoggerAdapter) -> None:
        """Initialize the mqtt client."""
        super().__init__(protocol=mqtt.MQTTv5)
        self._logger = logger
        self.update_client_id()

    def update_client_id(self):
        self._client_id = mqtt.base62(uuid.uuid4().int, padding=22)

    def ensure_loop_started(self) -> None:
        """Restart the loop if it is inactive."""
        if self._thread and self._thread.is_alive():
            _LOGGER.debug("Mqtt loop already running")
            return
        if self._thread:
            self._logger.debug("Stopping mqtt loop")
            super().loop_stop()
        super().loop_start()
        self._logger.debug("mqtt loop has been started")


class RoborockMqttClient(RoborockClient):
    def __init__(self, user_data: UserData, device_info: DeviceData, queue_timeout: int = 10) -> None:
        rriot = user_data.rriot
        if rriot is None:
            raise RoborockException("Got no rriot data from user_data")
        endpoint = base64.b64encode(Utils.md5(rriot.k.encode())[8:14]).decode()
        RoborockClient.__init__(self, endpoint, device_info, queue_timeout)
        self._logger = RoborockLoggerAdapter(device_info.device.name, _LOGGER)
        self._mqtt_user = rriot.u
        self._hashed_user = md5hex(self._mqtt_user + ":" + rriot.k)[2:10]
        url = urlparse(rriot.r.m)
        if not isinstance(url.hostname, str):
            raise RoborockException("Url parsing returned an invalid hostname")
        self._mqtt_host = str(url.hostname)
        self._mqtt_port = url.port
        self._mqtt_ssl = url.scheme == "ssl"
        self._mqtt_password = rriot.s
        self._hashed_password = md5hex(self._mqtt_password + ":" + rriot.k)[16:]
        self._endpoint = base64.b64encode(Utils.md5(rriot.k.encode())[8:14]).decode()
        self._waiting_queue: dict[int, RoborockFuture] = {}
        self._connection_lock = Lock()

        # Initialize the MQQ client library and set up callbacks
        client = MqttClient(self._logger)
        if self._mqtt_ssl:
            client.tls_set()
        client.username_pw_set(self._hashed_user, self._hashed_password)
        client.on_connect = self._mqtt_on_connect
        client.on_message = self._mqtt_on_message
        client.on_disconnect = self._mqtt_on_disconnect
        self._mqtt_client = client

    def _mqtt_on_connect(self, *args, **kwargs):
        _, __, ___, rc, ____ = args
        connection_queue = self._waiting_queue.get(CONNECT_REQUEST_ID)
        if rc != mqtt.MQTT_ERR_SUCCESS:
            message = f"Failed to connect ({mqtt.error_string(rc)})"
            self._logger.error(message)
            if connection_queue:
                connection_queue.resolve((None, VacuumError(message)))
            return
        self._logger.info(f"Connected to mqtt {self._mqtt_host}:{self._mqtt_port}")
        topic = f"rr/m/o/{self._mqtt_user}/{self._hashed_user}/{self.device_info.device.duid}"
        (result, _) = self._mqtt_client.subscribe(topic)
        if result != 0:
            message = f"Failed to subscribe ({mqtt.error_string(rc)})"
            self._logger.error(message)
            if connection_queue:
                connection_queue.resolve((None, VacuumError(message)))
            return
        self._logger.info(f"Subscribed to topic {topic}")
        if connection_queue:
            connection_queue.resolve((True, None))

    def _mqtt_on_message(self, *args, **kwargs):
        client, __, msg = args
        try:
            messages, _ = MessageParser.parse(msg.payload, local_key=self.device_info.device.local_key)
            super()._on_message_received(messages)
        except Exception as ex:
            self._logger.exception(ex)

    def _mqtt_on_disconnect(self, *args, **kwargs):
        _, __, rc, ___ = args
        try:
            exc = RoborockException(mqtt.error_string(rc)) if rc != mqtt.MQTT_ERR_SUCCESS else None
            super()._on_connection_lost(exc)
            if rc == mqtt.MQTT_ERR_PROTOCOL:
                self._mqtt_client.update_client_id()
            connection_queue = self._waiting_queue.get(DISCONNECT_REQUEST_ID)
            if connection_queue:
                connection_queue.resolve((True, None))
        except Exception as ex:
            self._logger.exception(ex)

    async def async_disconnect(self) -> None:
        async with self._connection_lock:
            if not self._mqtt_client.is_connected():
                return

            self._logger.info("Disconnecting from mqtt")
            disconnected_future = asyncio.ensure_future(self._async_response(DISCONNECT_REQUEST_ID))
            rc = self._mqtt_client.disconnect()

            if rc == mqtt.MQTT_ERR_NO_CONN:
                disconnected_future.cancel()
                return

            if rc != mqtt.MQTT_ERR_SUCCESS:
                disconnected_future.cancel()
                raise RoborockException(f"Failed to disconnect ({mqtt.error_string(rc)})")

            (_, err) = await disconnected_future
            if err:
                raise RoborockException(err) from err

    async def async_connect(self) -> None:
        async with self._connection_lock:
            if self._mqtt_client.is_connected():
                self._mqtt_client.ensure_loop_started()
                return

            if self._mqtt_port is None or self._mqtt_host is None:
                raise RoborockException("Mqtt information was not entered. Cannot connect.")

            self._logger.debug("Connecting to mqtt")
            connected_future = asyncio.ensure_future(self._async_response(CONNECT_REQUEST_ID))
            self._mqtt_client.connect(host=self._mqtt_host, port=self._mqtt_port, keepalive=KEEPALIVE)
            self._mqtt_client.ensure_loop_started()
            (_, err) = await connected_future
            if err:
                raise RoborockException(err) from err

    def _send_msg_raw(self, msg: bytes) -> None:
        info = self._mqtt_client.publish(
            f"rr/m/i/{self._mqtt_user}/{self._hashed_user}/{self.device_info.device.duid}", msg
        )
        if info.rc != mqtt.MQTT_ERR_SUCCESS:
            raise RoborockException(f"Failed to publish ({mqtt.error_string(info.rc)})")
