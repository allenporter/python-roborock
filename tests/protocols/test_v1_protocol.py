"""Tests for the v1 protocol message encoding and decoding."""

import json
import logging
import pathlib
from collections.abc import Generator
from unittest.mock import patch

import pytest
from freezegun import freeze_time
from syrupy import SnapshotAssertion

from roborock.data.containers import RoborockBase, UserData
from roborock.exceptions import RoborockException
from roborock.protocol import Utils
from roborock.protocols.v1_protocol import (
    RequestMessage,
    SecurityData,
    create_map_response_decoder,
    decode_rpc_response,
)
from roborock.roborock_message import RoborockMessage, RoborockMessageProtocol
from roborock.roborock_typing import RoborockCommand

from .. import mock_data

USER_DATA = UserData.from_dict(mock_data.USER_DATA)
TEST_REQUEST_ID = 44444
TEST_ENDPOINT = "87ItGWdb"
TEST_ENDPOINT_BYTES = TEST_ENDPOINT.encode()
SECURITY_DATA = SecurityData(
    endpoint=TEST_ENDPOINT,
    nonce=b"\x91\xbe\x10\xc9b+\x9d\x8a\xcdH*\x19\xf6\xfe\x81h",
)


TESTDATA_PATH = pathlib.Path("tests/protocols/testdata/v1_protocol/")
TESTDATA_FILES = list(TESTDATA_PATH.glob("*.json"))
TESTDATA_IDS = [x.stem for x in TESTDATA_FILES]


@pytest.fixture(autouse=True)
def fixed_time_fixture() -> Generator[None, None, None]:
    """Fixture to freeze time for predictable request IDs."""
    # Freeze time to a specific point so request IDs are predictable
    with freeze_time("2025-01-20T12:00:00"):
        yield


@pytest.fixture(name="test_request_id", autouse=True)
def request_id_fixture() -> Generator[int, None, None]:
    """Fixture to provide a fixed request ID."""
    with patch("roborock.protocols.v1_protocol.get_next_int", return_value=TEST_REQUEST_ID):
        yield TEST_REQUEST_ID


@pytest.mark.parametrize(
    ("command", "params", "expected"),
    [
        (
            RoborockCommand.GET_STATUS,
            None,
            b'{"dps":{"101":"{\\"id\\":44444,\\"method\\":\\"get_status\\",\\"params\\":[]}"},"t":1737374400}',
        )
    ],
)
def test_encode_local_payload(command, params, expected):
    """Test encoding of local payload for V1 commands."""
    message = RequestMessage(command, params).encode_message(RoborockMessageProtocol.GENERAL_REQUEST)
    assert isinstance(message, RoborockMessage)
    assert message.protocol == RoborockMessageProtocol.GENERAL_REQUEST
    assert message.payload == expected


@pytest.mark.parametrize(
    ("command", "params", "expected"),
    [
        (
            RoborockCommand.GET_STATUS,
            None,
            b'{"dps":{"101":"{\\"id\\":44444,\\"method\\":\\"get_status\\",\\"params\\":[],\\"security\\":{\\"endpoint\\":\\"87ItGWdb\\",\\"nonce\\":\\"91be10c9622b9d8acd482a19f6fe8168\\"}}"},"t":1737374400}',
        )
    ],
)
def test_encode_mqtt_payload(command, params, expected):
    """Test encoding of local payload for V1 commands."""
    request_message = RequestMessage(command, params=params)
    message = request_message.encode_message(RoborockMessageProtocol.RPC_REQUEST, SECURITY_DATA)
    assert isinstance(message, RoborockMessage)
    assert message.protocol == RoborockMessageProtocol.RPC_REQUEST
    assert message.payload == expected


@pytest.mark.parametrize(
    ("payload", "expected"),
    [
        (
            b'{"t":1652547161,"dps":{"102":"{\\"id\\":20005,\\"result\\":[{\\"msg_ver\\":2,\\"msg_seq\\":1072,\\"state\\":8,\\"battery\\":100,\\"clean_time\\":1041,\\"clean_area\\":37080000,\\"error_code\\":0,\\"map_present\\":1,\\"in_cleaning\\":0,\\"in_returning\\":0,\\"in_fresh_state\\":1,\\"lab_status\\":1,\\"water_box_status\\":0,\\"fan_power\\":103,\\"dnd_enabled\\":0,\\"map_status\\":3,\\"is_locating\\":0,\\"lock_status\\":0,\\"water_box_mode\\":202,\\"distance_off\\":0,\\"water_box_carriage_status\\":0,\\"mop_forbidden_enable\\":0,\\"unsave_map_reason\\":0,\\"unsave_map_flag\\":0}]}"}}',
            [
                {
                    "msg_ver": 2,
                    "msg_seq": 1072,
                    "state": 8,
                    "battery": 100,
                    "clean_time": 1041,
                    "clean_area": 37080000,
                    "error_code": 0,
                    "map_present": 1,
                    "in_cleaning": 0,
                    "in_returning": 0,
                    "in_fresh_state": 1,
                    "lab_status": 1,
                    "water_box_status": 0,
                    "fan_power": 103,
                    "dnd_enabled": 0,
                    "map_status": 3,
                    "is_locating": 0,
                    "lock_status": 0,
                    "water_box_mode": 202,
                    "distance_off": 0,
                    "water_box_carriage_status": 0,
                    "mop_forbidden_enable": 0,
                    "unsave_map_reason": 0,
                    "unsave_map_flag": 0,
                }
            ],
        ),
    ],
)
def test_decode_rpc_response(payload: bytes, expected: RoborockBase) -> None:
    """Test decoding a v1 RPC response protocol message."""
    # The values other than the payload are arbitrary
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.GENERAL_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )
    decoded_message = decode_rpc_response(message)
    assert decoded_message.request_id == 20005
    assert decoded_message.data == expected


@pytest.mark.parametrize("filename", TESTDATA_FILES, ids=TESTDATA_IDS)
def test_decode_rpc_payload(filename: str, snapshot: SnapshotAssertion) -> None:
    """Test decoding a v1 RPC response protocol message."""
    with open(filename, "rb") as f:
        payload = f.read()
    # The values other than the payload are arbitrary
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.GENERAL_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )
    decoded_message = decode_rpc_response(message)
    assert decoded_message.request_id == snapshot
    assert json.dumps(decoded_message.data, indent=2) == snapshot
    assert decoded_message.api_error is None


def test_create_map_response_decoder():
    """Test creating and using a map response decoder."""
    test_data = b"some map\n"
    compressed_data = (
        b"\x1f\x8b\x08\x08\xf9\x13\x99h\x00\x03foo\x00+\xce\xcfMU\xc8M,\xe0\x02\x00@\xdb\xc6\x1a\t\x00\x00\x00"
    )

    # Create header: endpoint(8) + padding(8) + request_id(2) + padding(6)
    # request_id = 44508 (0xaddc in little endian)
    header = TEST_ENDPOINT_BYTES + b"\x00" * 8 + b"\xdc\xad" + b"\x00" * 6
    encrypted_data = Utils.encrypt_cbc(compressed_data, SECURITY_DATA.nonce)
    payload = header + encrypted_data

    message = RoborockMessage(
        protocol=RoborockMessageProtocol.MAP_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )

    decoder = create_map_response_decoder(SECURITY_DATA)
    result = decoder(message)
    assert result is not None
    assert result.request_id == 44508
    assert result.data == test_data


def test_create_map_response_decoder_invalid_endpoint(caplog: pytest.LogCaptureFixture):
    """Test map response decoder with invalid endpoint."""
    caplog.set_level(logging.DEBUG)
    # Create header with wrong endpoint
    header = b"wrongend" + b"\x00" * 8 + b"\xdc\xad" + b"\x00" * 6
    payload = header + b"encrypted_data"

    message = RoborockMessage(
        protocol=RoborockMessageProtocol.MAP_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )

    decoder = create_map_response_decoder(SECURITY_DATA)
    assert decoder(message) is None
    assert "Received map response not requested by this device, ignoring." in caplog.text


def test_create_map_response_decoder_invalid_payload():
    """Test map response decoder with invalid payload."""
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.MAP_RESPONSE,
        payload=b"short",  # Too short payload
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )

    decoder = create_map_response_decoder(SECURITY_DATA)

    with pytest.raises(RoborockException, match="Invalid V1 map response format: missing payload"):
        decoder(message)


@pytest.mark.parametrize(
    ("payload", "expected_data", "expected_error"),
    [
        (
            b'{"t":1757883536,"dps":{"102":"{\\"id\\":20001,\\"result\\":\\"unknown_method\\"}"}}',
            {},
            "The method called is not recognized by the device.",
        ),
        (
            b'{"t":1757883536,"dps":{"102":"{\\"id\\":20001,\\"result\\":\\"other\\"}"}}',
            {},
            "Unexpected API Result",
        ),
    ],
)
def test_decode_result_with_error(payload: bytes, expected_data: dict[str, str], expected_error: str) -> None:
    """Test decoding a v1 RPC response protocol message."""
    # The values other than the payload are arbitrary
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.GENERAL_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )
    decoded_message = decode_rpc_response(message)
    assert decoded_message.request_id == 20001
    assert decoded_message.data == expected_data
    assert decoded_message.api_error
    assert expected_error in str(decoded_message.api_error)


def test_decode_no_request_id():
    """Test map response decoder without a request id is raised as an exception."""
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.GENERAL_RESPONSE,
        payload=b'{"t":1757883536,"dps":{"102":"{\\"result\\":\\"unknown_method\\"}"}}',
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )
    with pytest.raises(RoborockException, match="The method called is not recognized by the device"):
        decode_rpc_response(message)
