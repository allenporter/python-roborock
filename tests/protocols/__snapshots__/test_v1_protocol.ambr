# serializer version: 1
# name: test_decode_rpc_payload[app_get_init_status]
  20001
# ---
# name: test_decode_rpc_payload[app_get_init_status].1
  '''
  [
    {
      "local_info": {
        "name": "custom_A.03.0069_FCC",
        "bom": "A.03.0069",
        "location": "us",
        "language": "en",
        "wifiplan": "0x39",
        "timezone": "US/Pacific",
        "logserver": "awsusor0.fds.api.xiaomi.com",
        "featureset": 1
      },
      "feature_info": [
        111,
        112,
        113,
        114,
        115,
        116,
        117,
        118,
        119,
        120,
        122,
        123,
        124,
        125
      ],
      "new_feature_info": 633887780925447,
      "new_feature_info2": 8192,
      "new_feature_info_str": "0000000000002000",
      "status_info": {
        "state": 8,
        "battery": 100,
        "clean_time": 5610,
        "clean_area": 96490000,
        "error_code": 0,
        "in_cleaning": 0,
        "in_returning": 0,
        "in_fresh_state": 1,
        "lab_status": 1,
        "water_box_status": 0,
        "map_status": 3,
        "is_locating": 0,
        "lock_status": 0,
        "water_box_mode": 204,
        "distance_off": 0,
        "water_box_carriage_status": 0,
        "mop_forbidden_enable": 0
      }
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_clean_summary]
  20001
# ---
# name: test_decode_rpc_payload[get_clean_summary].1
  '''
  [
    1442559,
    24258125000,
    296,
    [
      1756848207,
      1754930385,
      1753203976,
      1752183435,
      1747427370,
      1746204046,
      1745601543,
      1744387080,
      1743528522,
      1742489154,
      1741022299,
      1740433682,
      1739902516,
      1738875106,
      1738864366,
      1738620067,
      1736873889,
      1736197544,
      1736121269,
      1734458038
    ]
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_consumeables]
  20001
# ---
# name: test_decode_rpc_payload[get_consumeables].1
  '''
  [
    {
      "main_brush_work_time": 879348,
      "side_brush_work_time": 707618,
      "filter_work_time": 738722,
      "filter_element_work_time": 0,
      "sensor_dirty_time": 455517
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_dnd]
  20002
# ---
# name: test_decode_rpc_payload[get_dnd].1
  '''
  [
    {
      "start_hour": 22,
      "start_minute": 0,
      "end_hour": 8,
      "end_minute": 0,
      "enabled": 1
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_last_clean_record]
  20003
# ---
# name: test_decode_rpc_payload[get_last_clean_record].1
  '''
  [
    [
      1738864366,
      1738868964,
      4358,
      81122500,
      0,
      0,
      1,
      1,
      21
    ]
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_multi_maps_list]
  20001
# ---
# name: test_decode_rpc_payload[get_multi_maps_list].1
  '''
  [
    {
      "max_multi_map": 1,
      "max_bak_map": 1,
      "multi_map_count": 1,
      "map_info": [
        {
          "mapFlag": 0,
          "add_time": 1747132930,
          "length": 0,
          "name": "",
          "bak_maps": [
            {
              "mapFlag": 4,
              "add_time": 1747132936
            }
          ]
        }
      ]
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_room_mapping2]
  20001
# ---
# name: test_decode_rpc_payload[get_room_mapping2].1
  '''
  [
    [
      16,
      "2537178",
      6
    ],
    [
      17,
      "2537175",
      14
    ],
    [
      18,
      "2537174",
      13
    ],
    [
      19,
      "2537176",
      14
    ],
    [
      20,
      "10655627",
      12
    ],
    [
      21,
      "2537145",
      2
    ],
    [
      22,
      "2537147",
      12
    ]
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_room_mapping3]
  20086
# ---
# name: test_decode_rpc_payload[get_room_mapping3].1
  '[]'
# ---
# name: test_decode_rpc_payload[get_room_mapping]
  20001
# ---
# name: test_decode_rpc_payload[get_room_mapping].1
  '''
  [
    [
      16,
      "3031886"
    ],
    [
      17,
      "3031880"
    ],
    [
      18,
      "3031883"
    ]
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_status]
  20001
# ---
# name: test_decode_rpc_payload[get_status].1
  '''
  [
    {
      "msg_ver": 2,
      "msg_seq": 515,
      "state": 8,
      "battery": 100,
      "clean_time": 5405,
      "clean_area": 91287500,
      "error_code": 0,
      "map_present": 1,
      "in_cleaning": 0,
      "in_returning": 0,
      "in_fresh_state": 1,
      "lab_status": 1,
      "water_box_status": 0,
      "fan_power": 106,
      "dnd_enabled": 1,
      "map_status": 3,
      "is_locating": 0,
      "lock_status": 0,
      "water_box_mode": 204,
      "distance_off": 0,
      "water_box_carriage_status": 0,
      "mop_forbidden_enable": 0,
      "unsave_map_reason": 4,
      "unsave_map_flag": 0
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_volume]
  20001
# ---
# name: test_decode_rpc_payload[get_volume].1
  '''
  [
    90
  ]
  '''
# ---
